### PR TITLE
wharf-api endpoints cleanup

### PR DESCRIPTION
**Pull request by applejag**
Restructing the wharf-api by changing the paths of the endpoints, adding and
renaming path parameters, and changing the request and response models around.

While we will keep backward compatibility for at least one major version, we
will be chaning so much that this needs a full major version bump.

This is not however not the full list of changes to be made for v5, but instead
a list of changes that will require us moving to v5. Other changes may be put
into the v5 bump that are eluded from this RFC.

## Rendered

Link to the markdown file here on GitHub. Example:
https://github.com/iver-wharf/rfcs/blob/rfc/wharf-api-v5/_published/0016-wharf-api-endpoints-cleanup.md

### CODE REVIEW

**Review by fredx30 — APPROVED** [review]
Content detail looks good to me. I think the title should be focused on the change of URL paths for organisation rather than the version number. As the version number itself does not reveal anything about intent.

**Comment by applejag** [comment]
> Content detail looks good to me. I think the title should be focused on the change of URL paths for organisation rather than the version number. As the version number itself does not reveal anything about intent.

Good point. I'll rename

File: _published/0016-wharf-api-endpoints-cleanup.md [hunk 1]
```diff
@@ -0,0 +1,849 @@
...
+
+- RFC PR: [iver-wharf/rfcs#{{page.pr_id}}](https://github.com/iver-wharf/rfcs/pulls/{{page.pr_id}})
+- Feature name: `{{page.rfc_feature_name}}`
+- Author: {{page.rfc_author_name}} ([@{{page.rfc_author_username}}](https://github.com/{{page.rfc_author_username}}))
+- Implementation issue: [{{page.impl_issue_repo}}#{{page.impl_issue_id}}](https://github.com/{{page.impl_issue_repo}}/issues/{{page.impl_issue_id}})
+- Implementation status: ![GitHub issue state](https://img.shields.io/github/issues/detail/state/{{page.impl_issue_repo}}/{{page.impl_issue_id}}?{{page.shields_io_query_params}})
+
+## Summary
+
+Restructing the wharf-api by changing the paths of the endpoints, adding and
```
Comment (Alexamakans) [inline _published/0016-wharf-api-endpoints-cleanup.md]:
```suggestion
Restructuring the wharf-api by changing the paths of the endpoints, adding and
```

File: _published/0016-wharf-api-endpoints-cleanup.md [hunk 2]
```diff
@@ -0,0 +1,849 @@
...
+renaming path parameters, and changing the request and response models around.
+
+While we will keep backward compatibility for at least one major version, we
+will be chaning so much that this needs a full major version bump.
```
Comment (Alexamakans) [inline _published/0016-wharf-api-endpoints-cleanup.md]:
```suggestion
will be changing so much that this needs a full major version bump.
```

File: _published/0016-wharf-api-endpoints-cleanup.md [hunk 3]
```diff
@@ -0,0 +1,849 @@
...
+
+This is not however not the full list of changes to be made for v5, but instead
+a list of changes that will require us moving to v5. Other changes may be put
+into the v5 bump that are eluded from this RFC.
```
Comment (Alexamakans) [inline _published/0016-wharf-api-endpoints-cleanup.md]: The current text, although I found it slightly hard to parse because "changes" and "eluded" are far from each other, is okay I think. Besides the double "not" that is.

nit: `v5 bump` -> `version bump`, either is fine but I think it reads better.

note: `eluded` means to escape, evade, dodge. I believe you were thinking of `elided`, i.e. `omitted`?

Rephrasings I like:
1
```suggestion
This is, however, not the full list of changes to be made for v5. Instead, it's
a list of changes that will require us moving to it. Changes not listed in
this RFC may be included in the version bump as well.
```
2
```suggestion
This is, however, not the full list of changes to be made for v5, but instead
those that will require us moving to it. Changes not included in this RFC
may be put into the version bump as well.
```

File: _published/0016-wharf-api-endpoints-cleanup.md [hunk 4]
```diff
@@ -0,0 +1,849 @@
...
+
+## Motivation
+
+We are lacking consistency in our API, and there are some questionable
+endpoints in there. Most annoying issues are:
+
+- Some endpoints are plural, some are singular.
+  Ex: `POST /project` vs `GET /projects`.
+
+- Some are misplaced and relies on request body instead of path/query
```
Comment (Alexamakans) [inline _published/0016-wharf-api-endpoints-cleanup.md]: 1
```suggestion
- Some are misplaced and rely on request body instead of path/query
```
2
```suggestion
- Some are misplaced, relying on the request body instead of path/query
```

File: _published/0016-wharf-api-endpoints-cleanup.md [hunk 5]
```diff
@@ -0,0 +1,849 @@
...
+  parameters. Ex: `PUT /branches` instead of `PUT /project/{projectId}/branch`.
+
+- Path parameters are all lowercased, while all other query parameters are
+  camelCased.
+
+- We use the database models in our request and response specs. This is a
+  security hole as it's easy to accidentally expose too much data to the
+  user. With separate models we would also make the Swagger documentation
+  more easily understood, where we could elude confusing specifications such as
```
Comment (Alexamakans) [inline _published/0016-wharf-api-endpoints-cleanup.md]: 1
```suggestion
  more easily understood, where we could elide confusing specifications such as
```
2
```suggestion
  more easily understood, where we could elide irrelevant specifications such as
```

File: _published/0016-wharf-api-endpoints-cleanup.md [hunk 6]
```diff
@@ -0,0 +1,849 @@
...
+  allowing to specify a project ID when creating a new project.
+
+## Explanation
+
+Moving from wharf-api v4 to v5, all endpoints are now solely in singular
+format, and some has been moved to a completely different subpath.
```
Comment (Alexamakans) [inline _published/0016-wharf-api-endpoints-cleanup.md]: 1
```suggestion
Moving from wharf-api v4 to v5, all endpoints are now in singular format, and
some have been moved to a completely different subpath.
```
2
```suggestion
Moving from wharf-api v4 to v5, all endpoints are now in singular format.
Some have also been moved to a completely different subpath.
```

File: _published/0016-wharf-api-endpoints-cleanup.md [hunk 7]
```diff
@@ -0,0 +1,849 @@
...
+
+Major changes:
+
+- POST endpoints no longer acts as "add or update", but instead solely as
+  "add".
+
+- PUT endpoints no longer acts as "add or update", but instead solely as
+  "update".
+
+- POST endpoints for searching has been removed, please refer to the GET
```
Comment (Alexamakans) [inline _published/0016-wharf-api-endpoints-cleanup.md]:
```suggestion
- POST endpoints for searching have been removed, please refer to the GET
```

File: _published/0016-wharf-api-endpoints-cleanup.md [hunk 8]
```diff
@@ -0,0 +1,849 @@
...
+  endpoints using query parameters instead of HTTP request body to query.
```
Comment (Alexamakans) [inline _published/0016-wharf-api-endpoints-cleanup.md]: nit:
```suggestion
  endpoints using query parameters instead of an HTTP request body to query.
```

File: _published/0016-wharf-api-endpoints-cleanup.md [hunk 9]
```diff
@@ -0,0 +1,849 @@
...
+
+  - This allows finer control over the search endpoints. Starting with v5, you
+    now have a range of `{field}Match` query parameters that does a
+    "soft match". Where instead of trying to match the result verbatim, the
```
Comment (Alexamakans) [inline _published/0016-wharf-api-endpoints-cleanup.md]: nit:
```suggestion
    "soft match". Where, instead of trying to match the result verbatim, the
```

File: _published/0016-wharf-api-endpoints-cleanup.md [hunk 10]
```diff
@@ -0,0 +1,849 @@
...
+    API will try a fuzzy or partial search for better human search results.
+
+- All endpoints without a path parameter declaring an ID are working with lists
+  and not single items. Some endpoints have been changed between v4 and v5 to
+  account for this, such as the `PUT /project -> PUT /project/{projectId}`
+  endpoint change.
+
+### Deprecation of endpoints
+
+All endpoints that has been moved or removed are **still fully functioning
```
Comment (Alexamakans) [inline _published/0016-wharf-api-endpoints-cleanup.md]:
```suggestion
All endpoints that have been moved or removed are **still fully functioning
```

File: _published/0016-wharf-api-endpoints-cleanup.md [hunk 11]
```diff
@@ -0,0 +1,849 @@
...
+throughout the entire wharf-api v5**. The old endpoints has been **marked as
+"Deprecated" in the Swagger/OpenAPI specification.**
+
+Word of caution is that they may very well be **removed in the next major
```
Comment (Alexamakans) [inline _published/0016-wharf-api-endpoints-cleanup.md]:
```suggestion
A word of caution, the deprecated endpoints may be **removed in the next major
```

File: _published/0016-wharf-api-endpoints-cleanup.md [hunk 12]
```diff
@@ -0,0 +1,849 @@
...
+version bump up to v6.** As a user of the wharf-api, you need to migrate your
+applications to use the new endpoints as specified in this RFC.
```
Comment (Alexamakans) [inline _published/0016-wharf-api-endpoints-cleanup.md]: The "endpoints specified in this RFC", or "the migration process specified in this RFC"?

If it's the former one then:
```suggestion
version bump up to v6.** As a user of the wharf-api, you need to migrate your
applications to use the new endpoints specified in this RFC.
```
Otherwise:
```suggestion
version bump up to v6.** As a user of the wharf-api, you need to migrate your
applications to use the new endpoints. You can do this using the method
specified in this RFC.
```
would be easier to understand I think.

File: _published/0016-wharf-api-endpoints-cleanup.md [hunk 13]
```diff
@@ -0,0 +1,849 @@
...
+of models inside the wharf-api:
+
+- `/pkg/models/database`: Database models, with [GORM](https://gorm.io/)
+  specific tags to declare table column names and SQL associations.
+
+  These are equivalent to the models from v4, though with any
+  [JSON](https://pkg.go.dev/encoding/json#Marshal) or other specific tags
+  trimmed away.
+
+- `/pkg/models/request`: HTTP request models sent by the client and received by
```
Comment (Alexamakans) [inline _published/0016-wharf-api-endpoints-cleanup.md]: nit: Consistency with the comma after "Database models" above.
```suggestion
- `/pkg/models/request`: HTTP request models, sent by the client and received by
```

File: _published/0016-wharf-api-endpoints-cleanup.md [hunk 14]
```diff
@@ -0,0 +1,849 @@
...
+  the server. Go fields shall be tagged with:
+
+  - [JSON](https://pkg.go.dev/encoding/json#Marshal) specific tags to declare
+    how they ought to be serialized
```
Comment (Alexamakans) [inline _published/0016-wharf-api-endpoints-cleanup.md]: "Ought to" has the connotation of there being a case where they are not serialized according to the tags.

If there is such a case, then just add the missing period.
```suggestion
    how they ought to be serialized.
```
Otherwise:
```suggestion
    how they will be serialized.
```
is better I think.

File: _published/0016-wharf-api-endpoints-cleanup.md [hunk 15]
```diff
@@ -0,0 +1,849 @@
...
+
+  - [Gin](https://gin-gonic.com/) specific tags such as which fields are
+    required or optional, and sensible defaults;
+
+  - and [Swaggo](https://github.com/swaggo/swag) specific tags for better
+    Swagger specification generation.
```
Comment (Alexamakans) [inline _published/0016-wharf-api-endpoints-cleanup.md]: nit: prefer not polluting markdown lists with grammatical dependencies between the elements
```suggestion
  - [Gin](https://gin-gonic.com/) specific tags such as which fields are
    required or optional, and sensible defaults.

  - [Swaggo](https://github.com/swaggo/swag) specific tags for better Swagger
    specification generation.
```

File: _published/0016-wharf-api-endpoints-cleanup.md [hunk 16]
```diff
@@ -0,0 +1,849 @@
...
+        Name string `json:"name" binding:"required" validate:"required" example:"github" enums:"github,gitlab,azuredevops"`
+        URL  string `json:"url" binding:"required" validate:"required" example:"https://api.github.com/"`
+    }
+    ```
+
+- `/pkg/models/response`: HTTP response models sent by the server and received
+  by the client. Go fields shall be tagged with:
+
+  - [JSON](https://pkg.go.dev/encoding/json#Marshal) specific tags to declare
+    how they ought to be serialized
```
Comment (Alexamakans) [inline _published/0016-wharf-api-endpoints-cleanup.md]: Same as the other one
1
```suggestion
  - [JSON](https://pkg.go.dev/encoding/json#Marshal) specific tags to declare
    how they ought to be serialized.
```
2
```suggestion
  - [JSON](https://pkg.go.dev/encoding/json#Marshal) specific tags to declare
    how they will be serialized.
```

File: _published/0016-wharf-api-endpoints-cleanup.md [hunk 17]
```diff
@@ -0,0 +1,849 @@
...
+
+  - and [Swaggo](https://github.com/swaggo/swag) specific tags for better
+    Swagger specification generation.
```
Comment (Alexamakans) [inline _published/0016-wharf-api-endpoints-cleanup.md]: nit: prefer not polluting markdown lists with grammatical dependencies between the elements
```suggestion
  - [Swaggo](https://github.com/swaggo/swag) specific tags for better Swagger
    specification generation.
```

File: _published/0016-wharf-api-endpoints-cleanup.md [hunk 18]
```diff
@@ -0,0 +1,849 @@
...
+
+  type Provider struct {
+      ID   int    `json:"id" example:"123"`
+      Name string `json:"name" example:"github" enums:"github,gitlab,azuredevops"`
+      URL  string `json:"url" example:"https://api.github.com/"`
+  }
+  ```
+
+Convertion between these are done explicitly on each usage to ensure no extra
+values are not leaked.
```
Comment (Alexamakans) [inline _published/0016-wharf-api-endpoints-cleanup.md]:
```suggestion
Conversion between these are done explicitly on each usage to ensure no extra
values are leaked.
```

File: _published/0016-wharf-api-endpoints-cleanup.md [hunk 19]
```diff
@@ -0,0 +1,849 @@
...
+
+#### New layout for models
+
+To reduce duplication in data and make things more intuitive for the user,
+endpoints now prioritizes the layout of the model instead of just using
```
Comment (Alexamakans) [inline _published/0016-wharf-api-endpoints-cleanup.md]:
```suggestion
endpoints now prioritize the layout of the model instead of just using
```

File: _published/0016-wharf-api-endpoints-cleanup.md [hunk 20]
```diff
@@ -0,0 +1,849 @@
...
+the database models.
+
+Some overarching design principles going from v4 to v5:
+
+- Refrain from relying on ID references in HTTP requests but instead rely on
+  query or path parameters.
```
Comment (Alexamakans) [inline _published/0016-wharf-api-endpoints-cleanup.md]: nit: Prefer shorter sentences with some pauses.
Not always easy to apply, but that's not the case here:
```suggestion
- Refrain from relying on ID references in HTTP requests. Instead, rely on
  query or path parameters.
```

Comment (Alexamakans) [reply _published/0016-wharf-api-endpoints-cleanup.md]: And perhaps a stricter word than "refrain" would be good?

File: _published/0016-wharf-api-endpoints-cleanup.md [hunk 21]
```diff
@@ -0,0 +1,849 @@
...
+
+- Never suggest that an ID can be provided when creating or altering an object
+  in POST or PUT requests.
```
Comment (Alexamakans) [inline _published/0016-wharf-api-endpoints-cleanup.md]: This confuses me a bit.

Is "Never suggest that an ID can be provided" referring to an ID in the request body?
If that's the case then I suggest some rephrasing.

Also, if that's the case, then right now it feels like it's saying kind of the same thing as the text above, but with the focus on a different aspect. That might just be me missing something though.

File: _published/0016-wharf-api-endpoints-cleanup.md [hunk 22]
```diff
@@ -0,0 +1,849 @@
...
+
+Some examples of these changes are the lot of PUT endpoints, such as
```
Comment (Alexamakans) [inline _published/0016-wharf-api-endpoints-cleanup.md]: nit: Since we're adhering to A.E. spelling, using "the lot of" which is more common in B.E. than A.E. feels off.
1
```suggestion
Some examples of these changes are the PUT endpoints, such as
```
2
```suggestion
Some examples of these changes are all of the PUT endpoints, such as
```
Alternatively,
```suggestion
The PUT endpoints now follow these principles, e.g.
```

File: _published/0016-wharf-api-endpoints-cleanup.md [hunk 23]
```diff
@@ -0,0 +1,849 @@
...
+
+Starting with wharf-api v5, all endpoints now have IDs. These are ignored in
+regular API usage, but when using code generation this comes in handy as the
+endpoint IDs can be used for functions/methods instead of some auto-generated
+names based on the path segments.
+
+Where previously the Swagger generated TypeScript method for
+`POST /project/{projectId}/build` *(formerly known as
+`POST /project/{projectid}/builds`)* was `projectProjectidBuildsGet()`, it will
+instead be `getProjectBranchList()`.
```
Comment (Alexamakans) [inline _published/0016-wharf-api-endpoints-cleanup.md]:
```suggestion
Where, previously, the Swagger generated TypeScript method for
`POST /project/{projectId}/build` *(formerly known as
`POST /project/{projectid}/builds`)* was `projectProjectidBuildsGet()`, it will
now instead be `getProjectBranchList()`.
```

File: _published/0016-wharf-api-endpoints-cleanup.md [hunk 24]
```diff
@@ -0,0 +1,849 @@
...
+
+### Renamed path parameters
+
+The path parameter has been transformed from lowercase to camelCase.
+This only affects code generators, such as the
+[Swagger Codegen](https://github.com/swagger-api/swagger-codegen), and has no
+implications in the actual HTTP requests nor responses.
```
Comment (Alexamakans) [inline _published/0016-wharf-api-endpoints-cleanup.md]:
```suggestion
implications in the actual HTTP requests nor the responses.
```

File: _published/0016-wharf-api-endpoints-cleanup.md [hunk 25]
```diff
@@ -0,0 +1,849 @@
...
+names based on the path segments.
+
+Where previously the Swagger generated TypeScript method for
+`POST /project/{projectId}/build` *(formerly known as
+`POST /project/{projectid}/builds`)* was `projectProjectidBuildsGet()`, it will
+instead be `getProjectBranchList()`.
+
+### Renamed path parameters
+
+The path parameter has been transformed from lowercase to camelCase.
```
Comment (Alexamakans) [inline _published/0016-wharf-api-endpoints-cleanup.md]: nit: `transformed` vs. `changed`
1
```suggestion
The path parameters have been transformed from lowercase to camelCase.
```
2
```suggestion
The path parameters have been changed from lowercase to camelCase.
```

File: _published/0016-wharf-api-endpoints-cleanup.md [hunk 26]
```diff
@@ -0,0 +1,849 @@
...
+  preference for the singular word forms, especially if we encounter some of
+  those that change drastically later, such as "person" vs "people".
+
+## Future possibilities
+
+<!--
+Does this lay groundwork for some future changes? If so, what?
+-->
+
+- As mentioned in the [#Compatability](#compatability), I have great plans for
```
Comment (Alexamakans) [inline _published/0016-wharf-api-endpoints-cleanup.md]:
```suggestion
- As mentioned in the [#Compatibility](#compatibility), I have great plans for
```

**Review by Alexamakans — COMMENTED** [review]
Not many comments regarding the design, will return with more comments regarding that later.

File: _published/0016-wharf-api-endpoints-cleanup.md (hunk 3 above)
Comment (applejag) [reply _published/0016-wharf-api-endpoints-cleanup.md]: I've a hard time with elide vs elude. I'll try use different words in the future when writing these documents

File: _published/0016-wharf-api-endpoints-cleanup.md (hunk 17 above)
Comment (applejag) [reply _published/0016-wharf-api-endpoints-cleanup.md]: I kind of like it, so you can read the list as if it was a sentence. So the list dots and indentation only become visual helpers, instead of the paragraph relying on them.

But I'm not :100: on this idea, so I'll stick with your suggestion as it feels safer

File: _published/0016-wharf-api-endpoints-cleanup.md (hunk 20 above)
Comment (applejag) [reply _published/0016-wharf-api-endpoints-cleanup.md]:
```suggestion
- Avoid relying on ID references in from the HTTP request body. Instead, rely on
  query or path parameters.
```

File: _published/0016-wharf-api-endpoints-cleanup.md (hunk 22 above)
Comment (applejag) [reply _published/0016-wharf-api-endpoints-cleanup.md]: ~~What's "A.E. spelling"?~~ Edit: Figured it out: American English vs Brittish English :)

File: _published/0016-wharf-api-endpoints-cleanup.md (hunk 26 above)
Comment (applejag) [reply _published/0016-wharf-api-endpoints-cleanup.md]: F*CK!

File: _published/0016-wharf-api-endpoints-cleanup.md (hunk 21 above)
Comment (applejag) [reply _published/0016-wharf-api-endpoints-cleanup.md]: The first bulletpoint is to use query parameters instead of request body.

This latter bulletpoint is to not specify in the model that the user may provide an ID in endpoints that create objects, such as `POST /project`

Will try to rephrase it to make it clearer

Comment (applejag) [reply _published/0016-wharf-api-endpoints-cleanup.md]: New:

> Some overarching design principles going from v4 to v5:
> 
> - Do not rely on ID references in the HTTP request body when targeting a specifc object. For example:
> 
>   - Good: `PUT /project/123`
> 
>     ```json
>     {
>       "name": "sample",
>       "description": "Sample project"
>     }
>     ```
> 
>   - Bad: `PUT /project`
> 
>     ```json
>     {
>       "projectId": 123,
>       "name": "sample",
>       "description": "Sample project"
>     }
>     ```
> 
> - Do not use a model in the Swagger documentation that suggests the user can provide an object ID for endpoints that creates objects. For example:
> 
>   - Good: `POST /project`
> 
>     ```json
>     {
>       "name": "sample",
>       "description": "Sample project"
>     }
>     ```
> 
>   - Bad: `POST /project`
> 
>     ```json
>     {
>       "projectId": 123,
>       "name": "sample",
>       "description": "Sample project"
>     }
>     ```

File: _published/0016-wharf-api-endpoints-cleanup.md (hunk 26 above)
Comment (Alexamakans) [reply _published/0016-wharf-api-endpoints-cleanup.md]: Hahahah 😂

File: _published/0016-wharf-api-endpoints-cleanup.md [hunk 27]
```diff
@@ -0,0 +1,888 @@
...
+  parameters. Ex: `PUT /branches` instead of `PUT /project/{projectId}/branch`.
+
+- Path parameters are all lowercased, while all other query parameters are
+  camelCased.
+
+- We use the database models in our request and response specs. This is a
+  security weakness as it induces a higher risk of exposing too much data to
+  the user. With separate models we would also make the Swagger documentation
+  easier to understand, where we could elide irrelevant specifications such as
+  allowing to specify a project ID when creating a new project.
```
Comment (Pikabanga) [inline _published/0016-wharf-api-endpoints-cleanup.md]: I think the strongest argument that you've missed is that this will hugely improve decoupling (i.e. easier to change the database model without changing the API model, making changes less viral) :)

File: _published/0016-wharf-api-endpoints-cleanup.md [hunk 28]
```diff
@@ -0,0 +1,888 @@
...
+Major changes:
+
+- POST endpoints no longer acts as "add or update", but instead solely as
+  "add".
+
+- PUT endpoints no longer acts as "add or update", but instead solely as
+  "update".
+
+- POST endpoints for searching have been removed, please refer to the GET
+  endpoints using query parameters instead of an HTTP request body to query.
```
Comment (Pikabanga) [inline _published/0016-wharf-api-endpoints-cleanup.md:88]: I would try to not be afraid of making exceptions to this rule if you need to allow for complex queries that become a pain to represent with query parameters (like nested conditions). There is a reason Elasticsearch queries are POST's.

File: _published/0016-wharf-api-endpoints-cleanup.md [hunk 29]
```diff
@@ -0,0 +1,888 @@
...
+This will break a lot of systems. Some rely on the "add or update" mechanics
+that I'm proposing to remove. But those systems should still be operable until
+v6, and IIRC that's the provider APIs that rely on them to simplify their
+own code.
+
+I have great plans for v6 with the [Hide providers behind API](https://iver-wharf.github.io/wharf-notes/hide-providers-behind-api)
+approach. This will render most of the "provider to API" communication obsolete
+as the API will rely on the provider's responses instead of requests.
+
+Again, it's of utter importance that the old endpoints works as before,
```
Comment (Pikabanga) [inline _published/0016-wharf-api-endpoints-cleanup.md]:
```suggestion
Again, it's of utmost importance that the old endpoints works as before,
```

File: _published/0016-wharf-api-endpoints-cleanup.md (hunk 28 above)
Comment (applejag) [reply _published/0016-wharf-api-endpoints-cleanup.md:88]: The query objects in elasticsearch are super useful, but I don't want to see something similar in Wharf because I think that would make it uncessecarily complex. The existing solution is kind of a hack IMO where it does a `.Where` clause in GORM on the entire JSON request body, which is what I wanted to avoid with this new rule.

But yea, if we need a more complex solution in the future where we want to allow fine-grained queries, then it would definetly be unwise to constrain ourselves with this rule.

File: _published/0016-wharf-api-endpoints-cleanup.md (hunk 27 above)
Comment (applejag) [reply _published/0016-wharf-api-endpoints-cleanup.md]: Gosh! Probably in the back of my mind as I was writing it, but thanks for reminding me!

**Review by Alexamakans — APPROVED** [review]

File: _published/0016-wharf-api-endpoints-cleanup.md (hunk 28 above)
Comment (applejag) [reply _published/0016-wharf-api-endpoints-cleanup.md:88]: I've added a note about this in the 'Alternatives' section
